<a href="https://colab.research.google.com/github/vedanshipathak/Plastic-Detection-Model/blob/main/Plastic_Detection_Model.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
!git clone https://github.com/pedropro/TACO.git

Cloning into 'TACO'...
remote: Enumerating objects: 740, done.
remote: Total 740 (delta 0), reused 0 (delta 0), pack-reused 740 (from 1)
Receiving objects: 100% (740/740), 107.77 MiB | 34.24 MiB/s, done.
Resolving deltas: 100% (498/498), done.


In [2]:
!cd TACO

In [3]:
!pip install pandas requests opencv-python pillow numpy

In [4]:
import pandas as pd

# Path to the CSV file
csv_path = "/content/TACO/data/all_image_urls.csv"

# Load the CSV file
urls_df = pd.read_csv(csv_path)

# Display the first few rows
print(urls_df.head())

  https://farm66.staticflickr.com/65535/33978196618_632623b4fc_z.jpg  \
0  https://farm66.staticflickr.com/65535/47803331...                   
1  https://farm66.staticflickr.com/65535/40888872...                   
2  https://farm66.staticflickr.com/65535/47803331...                   
3  https://farm66.staticflickr.com/65535/33978199...                   
4  https://farm66.staticflickr.com/65535/33978200...                   

  https://farm66.staticflickr.com/65535/33978196618_e30a59e0a8_o.png  
0  https://farm66.staticflickr.com/65535/47803331...                  
1  https://farm66.staticflickr.com/65535/40888872...                  
2  https://farm66.staticflickr.com/65535/47803331...                  
3  https://farm66.staticflickr.com/65535/33978199...                  
4  https://farm66.staticflickr.com/65535/33978200...                  


In [6]:
import pandas as pd

# Path to the CSV file
csv_path = "/content/TACO/data/all_image_urls.csv"

# Load the CSV file
urls_df = pd.read_csv(csv_path)

# Display the column names
print(urls_df.columns)

Index(['https://farm66.staticflickr.com/65535/33978196618_632623b4fc_z.jpg', 'https://farm66.staticflickr.com/65535/33978196618_e30a59e0a8_o.png'], dtype='object')


In [10]:
# Add column headers for the URL columns
urls_df.columns = ["Index"]  # Corrected column name

# Generate image IDs
urls_df["image_id"] = urls_df.index + 1  # Assign unique IDs starting from 1

# Display the updated DataFrame
print(urls_df.head())


ValueError: Length mismatch: Expected axis has 2 elements, new values have 1 elements

In [11]:
import pandas as pd

# Path to the CSV file
csv_path = "/content/TACO/data/all_image_urls.csv"

# Column names
column_names = ['Image_URL_Z', 'Image_URL_O']

# Load the CSV file with specified column names
urls_df = pd.read_csv(csv_path, names=column_names)

# Display the column names
print(urls_df.columns)


Index(['Image_URL_Z', 'Image_URL_O'], dtype='object')


In [12]:
import os

# Create a folder to store downloaded images
image_folder = "/content/TACO/data/images"
os.makedirs(image_folder, exist_ok=True)

print(f"Created folder: {image_folder}")

Created folder: /content/TACO/data/images


In [13]:
import requests
from PIL import Image
from io import BytesIO

# Function to download an image
def download_image(url, image_id):
    try:
        response = requests.get(url)
        response.raise_for_status()  # Check for errors
        image = Image.open(BytesIO(response.content))
        image.save(os.path.join(image_folder, f"{image_id}.jpg"))
        print(f"Downloaded image {image_id}")
    except Exception as e:
        print(f"Failed to download image {image_id}: {e}")

# Download all images
for index, row in urls_df.iterrows():
    image_id = index + 1  # Generate unique image IDs
    download_image(row["Image_URL_Z"], image_id)

Downloaded image 1
Downloaded image 2
Downloaded image 3
Downloaded image 4
Downloaded image 5
Downloaded image 6
Downloaded image 7
Downloaded image 8
Downloaded image 9
Downloaded image 10
Downloaded image 11
Downloaded image 12
Downloaded image 13
Downloaded image 14
Downloaded image 15
Downloaded image 16
Downloaded image 17
Downloaded image 18
Downloaded image 19
Downloaded image 20
Downloaded image 21
Downloaded image 22
Downloaded image 23
Downloaded image 24
Downloaded image 25
Downloaded image 26
Downloaded image 27
Downloaded image 28
Downloaded image 29
Downloaded image 30
Downloaded image 31
Downloaded image 32
Downloaded image 33
Downloaded image 34
Downloaded image 35
Downloaded image 36
Downloaded image 37
Downloaded image 38
Downloaded image 39
Downloaded image 40
Downloaded image 41
Downloaded image 42
Downloaded image 43
Downloaded image 44
Downloaded image 45
Downloaded image 46
Downloaded image 47
Downloaded image 48
Downloaded image 49
Downloaded image 50
Downloade

In [14]:
# List downloaded images
downloaded_images = os.listdir(image_folder)
print(f"Number of downloaded images: {len(downloaded_images)}")
print(downloaded_images[:5])  # Display the first 5 images

Number of downloaded images: 2596
['3420.jpg', '1212.jpg', '968.jpg', '3884.jpg', '3234.jpg']


In [15]:
# Download all images (skip rows with invalid URLs)
for index, row in urls_df.iterrows():
    if pd.notna(row["Image_URL_Z"]):  # Check if the URL is not NaN
        image_id = index + 1
        download_image(row["Image_URL_Z"], image_id)
    else:
        print(f"Skipped image {index + 1} (missing URL)")

Downloaded image 1
Downloaded image 2
Downloaded image 3
Downloaded image 4
Downloaded image 5
Downloaded image 6
Downloaded image 7
Downloaded image 8
Downloaded image 9
Downloaded image 10
Downloaded image 11
Downloaded image 12
Downloaded image 13
Downloaded image 14
Downloaded image 15
Downloaded image 16
Downloaded image 17
Downloaded image 18
Downloaded image 19
Downloaded image 20
Downloaded image 21
Downloaded image 22
Downloaded image 23
Downloaded image 24
Downloaded image 25
Downloaded image 26
Downloaded image 27
Downloaded image 28
Downloaded image 29
Downloaded image 30
Downloaded image 31
Downloaded image 32
Downloaded image 33
Downloaded image 34
Downloaded image 35
Downloaded image 36
Downloaded image 37
Downloaded image 38
Downloaded image 39
Downloaded image 40
Downloaded image 41
Downloaded image 42
Downloaded image 43
Downloaded image 44
Downloaded image 45
Downloaded image 46
Downloaded image 47
Downloaded image 48
Downloaded image 49
Downloaded image 50
Downloade

In [16]:
from google.colab import files

# Zip the images folder
!zip -r images.zip /content/TACO/data/images

# Download the zip file
files.download("images.zip")

  adding: content/TACO/data/images/ (stored 0%)
  adding: content/TACO/data/images/3420.jpg (deflated 0%)
  adding: content/TACO/data/images/1212.jpg (deflated 0%)
  adding: content/TACO/data/images/968.jpg (deflated 0%)
  adding: content/TACO/data/images/3884.jpg (deflated 0%)
  adding: content/TACO/data/images/3234.jpg (deflated 0%)
  adding: content/TACO/data/images/4441.jpg (deflated 0%)
  adding: content/TACO/data/images/1003.jpg (deflated 0%)
  adding: content/TACO/data/images/3496.jpg (deflated 0%)
  adding: content/TACO/data/images/4436.jpg (deflated 0%)
  adding: content/TACO/data/images/583.jpg (deflated 0%)
  adding: content/TACO/data/images/874.jpg (deflated 0%)
  adding: content/TACO/data/images/3382.jpg (deflated 0%)
  adding: content/TACO/data/images/4040.jpg (deflated 0%)
  adding: content/TACO/data/images/4445.jpg (deflated 0%)
  adding: content/TACO/data/images/117.jpg (deflated 0%)
  adding: content/TACO/data/images/860.jpg (deflated 0%)
  adding: content/TACO/data/i

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

 Extract Trash Objects
If you have the annotations file (annotations.json), use it to extract trash objects from the downloaded images.

In [17]:
import json
import numpy as np
import cv2

# Path to the annotations file
annotations_path = "/content/TACO/data/annotations.json"

# Load annotations
with open(annotations_path) as f:
    annotations = json.load(f)

# Function to extract trash objects
def extract_trash_objects(image_id, annotations):
    # Load image
    image_path = os.path.join(image_folder, f"{image_id}.jpg")
    image = Image.open(image_path).convert("RGBA")

    # Create a blank image for the trash object
    trash_object = Image.new("RGBA", image.size, (0, 0, 0, 0))

    # Iterate through annotations for the image
    for ann in annotations["annotations"]:
        if ann["image_id"] == image_id:
            # Get segmentation mask
            mask = np.zeros((image.size[1], image.size[0]), dtype=np.uint8)
            for segment in ann["segmentation"]:
                polygon = np.array(segment).reshape((-1, 2)).astype(np.int32)
                cv2.fillPoly(mask, [polygon], 1)

            # Apply mask to extract the trash object
            mask_image = Image.fromarray(mask * 255, "L")
            trash_object.paste(image, (0, 0), mask_image)

    return trash_object

# Example: Extract trash objects for the first image
image_id = 1  # Replace with the desired image ID
trash_object = extract_trash_objects(image_id, annotations)

# Save the extracted trash object
output_folder = "/content/TACO/data/extracted_trash_objects"
os.makedirs(output_folder, exist_ok=True)
trash_object.save(os.path.join(output_folder, f"trash_{image_id}.png"))

In [18]:
import os
import json
import numpy as np
from PIL import Image
import cv2

# Paths
annotations_path = "/content/TACO/data/annotations.json"
images_folder = "/content/TACO/data/images"
extracted_trash_folder = "/content/TACO/data/extracted_trash_objects"
os.makedirs(extracted_trash_folder, exist_ok=True)

# Load annotations
with open(annotations_path) as f:
    annotations = json.load(f)

# Function to extract trash objects
def extract_trash_objects(image_id, annotations):
    # Load image
    image_path = os.path.join(images_folder, f"{image_id}.jpg")
    if not os.path.exists(image_path):
        print(f"Image {image_id}.jpg not found. Skipping.")
        return None

    image = Image.open(image_path).convert("RGBA")

    # Create a blank image for the trash object
    trash_object = Image.new("RGBA", image.size, (0, 0, 0, 0))

    # Iterate through annotations for the image
    for ann in annotations["annotations"]:
        if ann["image_id"] == image_id:
            # Get segmentation mask
            mask = np.zeros((image.size[1], image.size[0]), dtype=np.uint8)
            for segment in ann["segmentation"]:
                polygon = np.array(segment).reshape((-1, 2)).astype(np.int32)
                cv2.fillPoly(mask, [polygon], 1)

            # Apply mask to extract the trash object
            mask_image = Image.fromarray(mask * 255, "L")
            trash_object.paste(image, (0, 0), mask_image)

    return trash_object

# Extract trash objects for all images
for image_info in annotations["images"]:
    image_id = image_info["id"]
    trash_object = extract_trash_objects(image_id, annotations)
    if trash_object:
        trash_object.save(os.path.join(extracted_trash_folder, f"trash_{image_id}.png"))
        print(f"Extracted trash object for image {image_id}")

Image 0.jpg not found. Skipping.
Extracted trash object for image 1
Extracted trash object for image 2
Extracted trash object for image 3
Extracted trash object for image 4
Extracted trash object for image 5
Extracted trash object for image 6
Extracted trash object for image 7
Extracted trash object for image 8
Extracted trash object for image 9
Extracted trash object for image 10
Extracted trash object for image 11
Extracted trash object for image 12
Extracted trash object for image 13
Extracted trash object for image 14
Extracted trash object for image 15
Extracted trash object for image 16
Extracted trash object for image 17
Extracted trash object for image 18
Extracted trash object for image 19
Extracted trash object for image 20
Extracted trash object for image 21
Extracted trash object for image 22
Extracted trash object for image 23
Extracted trash object for image 24
Extracted trash object for image 25
Extracted trash object for image 26
Extracted trash object for image 27
Extr

In [19]:
# List extracted trash objects
extracted_images = os.listdir(extracted_trash_folder)
print(f"Number of extracted trash objects: {len(extracted_images)}")
print(extracted_images[:5])  # Display the first 5 extracted images

Number of extracted trash objects: 1216
['trash_285.png', 'trash_615.png', 'trash_57.png', 'trash_1015.png', 'trash_185.png']


In [20]:
# Extract trash objects for all images (with error handling)
for image_info in annotations["images"]:
    image_id = image_info["id"]
    try:
        trash_object = extract_trash_objects(image_id, annotations)
        if trash_object:
            trash_object.save(os.path.join(extracted_trash_folder, f"trash_{image_id}.png"))
            print(f"Extracted trash object for image {image_id}")
    except Exception as e:
        print(f"Error processing image {image_id}: {e}")

Image 0.jpg not found. Skipping.
Extracted trash object for image 1
Extracted trash object for image 2
Extracted trash object for image 3
Extracted trash object for image 4
Extracted trash object for image 5
Extracted trash object for image 6
Extracted trash object for image 7
Extracted trash object for image 8
Extracted trash object for image 9
Extracted trash object for image 10
Extracted trash object for image 11
Extracted trash object for image 12
Extracted trash object for image 13
Extracted trash object for image 14
Extracted trash object for image 15
Extracted trash object for image 16
Extracted trash object for image 17
Extracted trash object for image 18
Extracted trash object for image 19
Extracted trash object for image 20
Extracted trash object for image 21
Extracted trash object for image 22
Extracted trash object for image 23
Extracted trash object for image 24
Extracted trash object for image 25
Extracted trash object for image 26
Extracted trash object for image 27
Extr

In [21]:
from collections import Counter

# List all files in the folder
extracted_images = os.listdir(extracted_trash_folder)

# Check for duplicate filenames
duplicates = [filename for filename, count in Counter(extracted_images).items() if count > 1]

if duplicates:
    print("Duplicate filenames found:", duplicates)
else:
    print("All filenames are unique.")

All filenames are unique.


TO ORGANIZE THE DATASET

In [25]:
import os
import shutil
from sklearn.model_selection import train_test_split

# Paths
extracted_trash_folder = "/content/TACO/data/extracted_trash_objects"
dataset_folder = "/content/plastic_waste_dataset"
os.makedirs(dataset_folder, exist_ok=True)

# Create subfolders
for split in ["train", "val", "test"]:
    os.makedirs(os.path.join(dataset_folder, split, "images"), exist_ok=True)
    os.makedirs(os.path.join(dataset_folder, split, "masks"), exist_ok=True)

# List all extracted trash objects
extracted_images = [f for f in os.listdir(extracted_trash_folder) if f.endswith(".png")]

# Split the dataset (80% train, 20% temp)
train_images, temp_images = train_test_split(extracted_images, test_size=0.2, random_state=42)

# Split the temp set into validation and test (50% val, 50% test)
val_images, test_images = train_test_split(temp_images, test_size=0.5, random_state=42)

# Function to copy files to the respective folders
def copy_files(file_list, split):
    for file in file_list:
        shutil.copy(
            os.path.join(extracted_trash_folder, file),
            os.path.join(dataset_folder, split, "images", file)
        )
        # If you have masks, copy them as well
        # shutil.copy(
        #     os.path.join(extracted_trash_folder, "masks", file),
        #     os.path.join(dataset_folder, split, "masks", file)
        # )

# Copy files to train, val, and test folders
copy_files(train_images, "train")
copy_files(val_images, "val")
copy_files(test_images, "test")

print("Dataset organized successfully!")

Dataset organized successfully!


If the total number of images is 1216 (972 + 122 + 122), the percentages are:

This matches the desired 80-10-10 split.

In [26]:
print(f"Train images: {len(train_images)}")
print(f"Validation images: {len(val_images)}")
print(f"Test images: {len(test_images)}")

Train images: 972
Validation images: 122
Test images: 122


In [27]:
# Function to copy masks
def copy_masks(file_list, split):
    for file in file_list:
        shutil.copy(
            os.path.join(extracted_trash_masks, file),  # Path to masks
            os.path.join(dataset_folder, split, "masks", file)
        )

# Copy masks to train, val, and test folders
copy_masks(train_images, "train")
copy_masks(val_images, "val")
copy_masks(test_images, "test")

NameError: name 'extracted_trash_masks' is not defined